AX: Implement support for serving AXVisibleCharacterRange off the main-thread

twilco opened the pull request:
#### 87a311bda9faad38307d5decfae9e18cfa386752
<pre>
AX: Implement support for serving AXVisibleCharacterRange off the main-thread
<a href="https://bugs.webkit.org/show_bug.cgi?id=294619">https://bugs.webkit.org/show_bug.cgi?id=294619</a>
<a href="https://rdar.apple.com/153652827">rdar://153652827</a>

Reviewed by Joshua Hoffman.

With this commit, we implement support for serving AXVisibleCharacterRange off the main-thread. The high level approach is:

  1. Accessibility paints already paint all InlineIterator::TextBoxes that are visible. When this happens, keep track
     of the line indices of the painted text boxes, and associate that with a RenderText, which in turn is associated
     with an accessibility object.

  2. Cache the first painted line and last painted line index for each RenderText in a HashMap (m_mostRecentlyPaintedText),
     and sync this HashMap to the accessibility thread.

  3. Implement AXIsolatedObject::visibleCharacterRange using this HashMap. This is trivial for leaf nodes. For non-leaf
     nodes, we iterate over the descendants until we find the first painted one. The `location` of the visible NSRange
     is the text of all descendants (which are unpainted) up to the first painted one. The `length` of the visible range
     is then the length from the first painted line, to the last painted line of the last painted descendant.

This algorithm produces very different results from the main-thread visible character range implementation, which is
based on VisiblePosition line-by-line iteration to fit an objects element rect into the viewport rect. In general, the
new algorithm seems much more accurate, and this can be seen in the layout tests modified in this patch (I left code
comments in each test).

This patch also stops outputting AXVisibleCharacterRange in tests that dump every attribute for every element. Because
the algorithms produce slightly different results in ways that are not clearly relevant to any user-facing behavior,
not outputting AXVisibleCharacterRange in these tests allows us to keep the tests running in both configurations. Some
tests have also been re-written to include better test corpus text — logging &quot;1: Lineone lineone&quot;, &quot;2: Linetwo linetwo&quot;
rather than the word &quot;test&quot; thousands of times, which is harder to debug with the --show-window test runner flag.

* LayoutTests/accessibility/image-link-expected.txt:
* LayoutTests/accessibility/image-map2-expected.txt:
* LayoutTests/accessibility/internal-link-anchors2-expected.txt:
* LayoutTests/accessibility/mac/document-attributes-expected.txt:
* LayoutTests/accessibility/mac/document-links-expected.txt:
* LayoutTests/accessibility/mac/visible-character-range-width-changes-expected.txt: Added.
* LayoutTests/accessibility/mac/visible-character-range-width-changes.html: Added.
* LayoutTests/accessibility/mixed-contenteditable-double-br-visible-character-range-hang-expected.txt:
* LayoutTests/accessibility/mixed-contenteditable-double-br-visible-character-range-hang.html:
* LayoutTests/accessibility/mixed-contenteditable-visible-character-range-hang-expected.txt:
* LayoutTests/accessibility/mixed-contenteditable-visible-character-range-hang.html:
* LayoutTests/accessibility/table-attributes-expected.txt:
* LayoutTests/accessibility/table-multiple-tbodies-expected.txt:
* LayoutTests/accessibility/table-one-cell-expected.txt:
* LayoutTests/accessibility/table-thead-tfoot-expected.txt:
* LayoutTests/accessibility/table-with-rules-expected.txt:
* LayoutTests/accessibility/transformed-element-expected.txt:
* LayoutTests/accessibility/visible-character-range-basic-expected.txt: Added.
* LayoutTests/accessibility/visible-character-range-basic.html:
* LayoutTests/accessibility/visible-character-range-height-changes-expected.txt: Added.
* LayoutTests/accessibility/visible-character-range-height-changes.html:
* LayoutTests/accessibility/visible-character-range-scrolling.html:
* LayoutTests/accessibility/visible-character-range-vertical-rl-expected.txt:
* LayoutTests/accessibility/visible-character-range-vertical-rl.html:
* LayoutTests/accessibility/visible-character-range-width-changes.html: Removed.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/ios/accessibility/mixed-contenteditable-double-br-visible-character-range-hang-expected.txt: Removed.
* LayoutTests/platform/ios/accessibility/mixed-contenteditable-visible-character-range-hang-expected.txt: Removed.
* LayoutTests/platform/ios/accessibility/visible-character-range-basic-expected.txt: Removed.
* LayoutTests/platform/ios/accessibility/visible-character-range-height-changes-expected.txt: Removed.
* LayoutTests/platform/ios/accessibility/visible-character-range-width-changes-expected.txt: Removed.
* LayoutTests/platform/mac/accessibility/lists-expected.txt:
* LayoutTests/platform/mac/accessibility/math-multiscript-attributes-expected.txt:
* LayoutTests/platform/mac/accessibility/visible-character-range-basic-expected.txt: Removed.
* LayoutTests/platform/mac/accessibility/visible-character-range-height-changes-expected.txt: Removed.
* LayoutTests/platform/mac/accessibility/visible-character-range-scrolling-expected.txt:
* LayoutTests/platform/mac/accessibility/visible-character-range-width-changes-expected.txt: Removed.
* LayoutTests/resources/accessibility-helper.js:
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::LineRange::LineRange):
(WebCore::LineRange::debugDescription const):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::mostRecentlyPaintedText):
(WebCore::AXObjectCache::onAccessibilityPaintStarted):
(WebCore::AXObjectCache::onAccessibilityPaintFinished):
(WebCore::AXObjectCache::onPaint):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AXTextRun.h:
(WebCore::AXTextRun::characterAt const):
(WebCore::AXTextRun::textLength const):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::previousLineStartBoundaryPoints const): Deleted.
(WebCore::AccessibilityObject::lastBoundaryPointContainedInRect const): Deleted.
(WebCore::textStartPoint): Deleted.
(WebCore::textEndPoint): Deleted.
(WebCore::AccessibilityObject::boundaryPointsContainedInRect const): Deleted.
(WebCore::AccessibilityObject::visibleCharacterRange const): Deleted.
(WebCore::AccessibilityObject::visibleCharacterRangeInternal const): Deleted.
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/cocoa/AccessibilityObjectCocoa.mm:
(WebCore::textStartPoint):
(WebCore::textEndPoint):
(WebCore::AccessibilityObject::previousLineStartBoundaryPoints const):
(WebCore::AccessibilityObject::lastBoundaryPointContainedInRect const):
(WebCore::AccessibilityObject::boundaryPointsContainedInRect const):
(WebCore::AccessibilityObject::visibleCharacterRange const):
(WebCore::AccessibilityObject::visibleCharacterRangeInternal const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::visibleCharacterRange const): Deleted.
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::relatedObjectIDsFor):
(WebCore::AXIsolatedTree::applyPendingChanges):
(WebCore::AXIsolatedTree::sortedNonRootWebAreasDidChange):
(WebCore::AXIsolatedTree::processQueuedNodeUpdates):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
(WebCore::AXIsolatedTree::markMostRecentlyPaintedTextDirty):
(WebCore::AXIsolatedTree::mostRecentlyPaintedText const):
* Source/WebCore/accessibility/isolatedtree/mac/AXIsolatedObjectMac.mm:
(WebCore::AXIsolatedObject::visibleCharacterRange const):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperBase.mm:
(-[WebAccessibilityObjectWrapperBase accessibilityVisibleCharacterRange]):
* Source/WebCore/dom/BoundaryPoint.cpp:
(WebCore::BoundaryPoint::debugDescription const):
* Source/WebCore/dom/BoundaryPoint.h:
* Source/WebCore/dom/SimpleRange.cpp:
(WebCore::SimpleRange::debugDescription const):
* Source/WebCore/dom/SimpleRange.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::doAfterUpdateRendering):
* Source/WebCore/rendering/AccessibilityRegionContext.cpp:
(WebCore::AccessibilityRegionContext::~AccessibilityRegionContext):
(WebCore::AccessibilityRegionContext::takeBounds):
* Source/WebCore/rendering/AccessibilityRegionContext.h:
* Source/WebCore/rendering/TextBoxPainter.cpp:
(WebCore::TextBoxPainter::paint):
* Tools/WebKitTestRunner/InjectedBundle/mac/AccessibilityUIElementMac.mm:
(WTR::AccessibilityUIElement::allAttributes):

Canonical link: <a href="https://commits.webkit.org/296359@main">https://commits.webkit.org/296359@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e53ec8b73a193799caf21a5bc5106acd2a2c2285

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108283 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27944 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18366 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113492 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/58725 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110246 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28634 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36493 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82213 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/58725 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111231 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22692 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97541 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62649 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22106 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15679 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58225 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92064 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15735 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116614 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35338 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/26020 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91247 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/35710 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93818 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91048 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23201 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35931 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13700 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/31096 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35239 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/40780 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34959 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38313 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36637 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

minorninth commented on LayoutTests/accessibility/mac/visible-character-range-width-changes.html:
```diff
@@ -0,0 +1,81 @@
+        output += await verifyVisibleRange(text, {width: 270, height: 500, scrollTop: 198}, ["{135, 551}", "{137, 550}"]);
+        output += await verifyVisibleRange(text, {width: 320, height: 500, scrollTop: 198}, ["{135, 551}", "{115, 572}"]);
+
+        document.getElementById("text-div").style.visibility = "hidden";
```
I'm curious, if you use display:none instead can you eliminate the extra whitespace at the end of the results?

twilco replied:
Yes! Done in the latest commit.

minorninth commented on Source/WebCore/accessibility/AXCoreObject.h:
```diff
@@ -678,6 +678,25 @@ struct TextUnderElementMode {
 
+struct LineRange {
+    unsigned startLineIndex;
+    unsigned endLineIndex;
```
For any range data type I think it's helpful to include a comment saying whether the end index is inclusive or exclusive

twilco replied:
Good idea! Done in latest commit.

minorninth commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -2067,7 +2067,48 @@ void AXObjectCache::onStyleChange(Element& element, OptionSet<Style::Change> cha
+
+    for (auto iterator = m_mostRecentlyPaintedText.begin(); iterator != m_mostRecentlyPaintedText.end(); ++iterator) {
+        const auto& renderText = iterator->key;
+        // FIXME: Use InlineIteratorLogicalOrderTraversal instead. Otherwise we'll do the wrong thing for mixed direction
```
Is there a radar for this?

twilco replied:
I just filed:

https://bugs.webkit.org/show_bug.cgi?id=294632

And added a link to this bug in code comments in both places that will need changes.

twilco committed:
AX: Implement support for serving AXVisibleCharacterRange off the main-thread
https://bugs.webkit.org/show_bug.cgi?id=294619
rdar://153652827

Reviewed by Joshua Hoffman.

With this commit, we implement support for serving AXVisibleCharacterRange off the main-thread. The high level approach is:

  1. Accessibility paints already paint all InlineIterator::TextBoxes that are visible. When this happens, keep track
     of the line indices of the painted text boxes, and associate that with a RenderText, which in turn is associated
     with an accessibility object.

  2. Cache the first painted line and last painted line index for each RenderText in a HashMap (m_mostRecentlyPaintedText),
     and sync this HashMap to the accessibility thread.

  3. Implement AXIsolatedObject::visibleCharacterRange using this HashMap. This is trivial for leaf nodes. For non-leaf
     nodes, we iterate over the descendants until we find the first painted one. The `location` of the visible NSRange
     is the text of all descendants (which are unpainted) up to the first painted one. The `length` of the visible range
     is then the length from the first painted line, to the last painted line of the last painted descendant.

This algorithm produces very different results from the main-thread visible character range implementation, which is
based on VisiblePosition line-by-line iteration to fit an objects element rect into the viewport rect. In general, the
new algorithm seems much more accurate, and this can be seen in the layout tests modified in this patch (I left code
comments in each test).

This patch also stops outputting AXVisibleCharacterRange in tests that dump every attribute for every element. Because
the algorithms produce slightly different results in ways that are not clearly relevant to any user-facing behavior,
not outputting AXVisibleCharacterRange in these tests allows us to keep the tests running in both configurations. Some
tests have also been re-written to include better test corpus text — logging "1: Lineone lineone", "2: Linetwo linetwo"
rather than the word "test" thousands of times, which is harder to debug with the --show-window test runner flag.

* LayoutTests/accessibility/image-link-expected.txt:
* LayoutTests/accessibility/image-map2-expected.txt:
* LayoutTests/accessibility/internal-link-anchors2-expected.txt:
* LayoutTests/accessibility/mac/document-attributes-expected.txt:
* LayoutTests/accessibility/mac/document-links-expected.txt:
* LayoutTests/accessibility/mac/visible-character-range-width-changes-expected.txt: Added.
* LayoutTests/accessibility/mac/visible-character-range-width-changes.html: Added.
* LayoutTests/accessibility/mixed-contenteditable-double-br-visible-character-range-hang-expected.txt:
* LayoutTests/accessibility/mixed-contenteditable-double-br-visible-character-range-hang.html:
* LayoutTests/accessibility/mixed-contenteditable-visible-character-range-hang-expected.txt:
* LayoutTests/accessibility/mixed-contenteditable-visible-character-range-hang.html:
* LayoutTests/accessibility/table-attributes-expected.txt:
* LayoutTests/accessibility/table-multiple-tbodies-expected.txt:
* LayoutTests/accessibility/table-one-cell-expected.txt:
* LayoutTests/accessibility/table-thead-tfoot-expected.txt:
* LayoutTests/accessibility/table-with-rules-expected.txt:
* LayoutTests/accessibility/transformed-element-expected.txt:
* LayoutTests/accessibility/visible-character-range-basic-expected.txt: Added.
* LayoutTests/accessibility/visible-character-range-basic.html:
* LayoutTests/accessibility/visible-character-range-height-changes-expected.txt: Added.
* LayoutTests/accessibility/visible-character-range-height-changes.html:
* LayoutTests/accessibility/visible-character-range-scrolling.html:
* LayoutTests/accessibility/visible-character-range-vertical-rl-expected.txt:
* LayoutTests/accessibility/visible-character-range-vertical-rl.html:
* LayoutTests/accessibility/visible-character-range-width-changes.html: Removed.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/ios/accessibility/mixed-contenteditable-double-br-visible-character-range-hang-expected.txt: Removed.
* LayoutTests/platform/ios/accessibility/mixed-contenteditable-visible-character-range-hang-expected.txt: Removed.
* LayoutTests/platform/ios/accessibility/visible-character-range-basic-expected.txt: Removed.
* LayoutTests/platform/ios/accessibility/visible-character-range-height-changes-expected.txt: Removed.
* LayoutTests/platform/ios/accessibility/visible-character-range-width-changes-expected.txt: Removed.
* LayoutTests/platform/mac/accessibility/lists-expected.txt:
* LayoutTests/platform/mac/accessibility/math-multiscript-attributes-expected.txt:
* LayoutTests/platform/mac/accessibility/visible-character-range-basic-expected.txt: Removed.
* LayoutTests/platform/mac/accessibility/visible-character-range-height-changes-expected.txt: Removed.
* LayoutTests/platform/mac/accessibility/visible-character-range-scrolling-expected.txt:
* LayoutTests/platform/mac/accessibility/visible-character-range-width-changes-expected.txt: Removed.
* LayoutTests/resources/accessibility-helper.js:
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::LineRange::LineRange):
(WebCore::LineRange::debugDescription const):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::mostRecentlyPaintedText):
(WebCore::AXObjectCache::onAccessibilityPaintStarted):
(WebCore::AXObjectCache::onAccessibilityPaintFinished):
(WebCore::AXObjectCache::onPaint):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AXTextRun.h:
(WebCore::AXTextRun::characterAt const):
(WebCore::AXTextRun::textLength const):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::previousLineStartBoundaryPoints const): Deleted.
(WebCore::AccessibilityObject::lastBoundaryPointContainedInRect const): Deleted.
(WebCore::textStartPoint): Deleted.
(WebCore::textEndPoint): Deleted.
(WebCore::AccessibilityObject::boundaryPointsContainedInRect const): Deleted.
(WebCore::AccessibilityObject::visibleCharacterRange const): Deleted.
(WebCore::AccessibilityObject::visibleCharacterRangeInternal const): Deleted.
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/cocoa/AccessibilityObjectCocoa.mm:
(WebCore::textStartPoint):
(WebCore::textEndPoint):
(WebCore::AccessibilityObject::previousLineStartBoundaryPoints const):
(WebCore::AccessibilityObject::lastBoundaryPointContainedInRect const):
(WebCore::AccessibilityObject::boundaryPointsContainedInRect const):
(WebCore::AccessibilityObject::visibleCharacterRange const):
(WebCore::AccessibilityObject::visibleCharacterRangeInternal const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::visibleCharacterRange const): Deleted.
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::relatedObjectIDsFor):
(WebCore::AXIsolatedTree::applyPendingChanges):
(WebCore::AXIsolatedTree::sortedNonRootWebAreasDidChange):
(WebCore::AXIsolatedTree::processQueuedNodeUpdates):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
(WebCore::AXIsolatedTree::markMostRecentlyPaintedTextDirty):
(WebCore::AXIsolatedTree::mostRecentlyPaintedText const):
* Source/WebCore/accessibility/isolatedtree/mac/AXIsolatedObjectMac.mm:
(WebCore::AXIsolatedObject::visibleCharacterRange const):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperBase.mm:
(-[WebAccessibilityObjectWrapperBase accessibilityVisibleCharacterRange]):
* Source/WebCore/dom/BoundaryPoint.cpp:
(WebCore::BoundaryPoint::debugDescription const):
* Source/WebCore/dom/BoundaryPoint.h:
* Source/WebCore/dom/SimpleRange.cpp:
(WebCore::SimpleRange::debugDescription const):
* Source/WebCore/dom/SimpleRange.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::doAfterUpdateRendering):
* Source/WebCore/rendering/AccessibilityRegionContext.cpp:
(WebCore::AccessibilityRegionContext::~AccessibilityRegionContext):
(WebCore::AccessibilityRegionContext::takeBounds):
* Source/WebCore/rendering/AccessibilityRegionContext.h:
* Source/WebCore/rendering/TextBoxPainter.cpp:
(WebCore::TextBoxPainter::paint):
* Tools/WebKitTestRunner/InjectedBundle/mac/AccessibilityUIElementMac.mm:
(WTR::AccessibilityUIElement::allAttributes):

Canonical link: https://commits.webkit.org/296359@main
